File: app/src/main/java/com/healios/dreams/ui/permissions/PermissionRecyclerViewAdapter.java
package com.healios.dreams.ui.permissions;

import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.healios.dreams.model.PermissionModel;

import java.util.List;

public class PermissionRecyclerViewAdapter extends RecyclerView.Adapter<PermissionRecyclerViewAdapter.PermissionViewHolder> {

    private List<PermissionModel> permissionList;

    private PermissionRecyclerViewListener listener;

    public PermissionRecyclerViewAdapter(PermissionRecyclerViewListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public PermissionViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new PermissionViewHolder(new PermissionItemView(parent.getContext()), listener);
    }

    @Override
    public void onBindViewHolder(@NonNull PermissionViewHolder holder, int position) {
        PermissionModel permissionModel = permissionList.get(position);
        holder.bindModel(permissionModel);
    }

    @Override
    public int getItemCount() {
        if (permissionList != null) {
            return permissionList.size();
        }
        return 0;
    }

    //region: Setters
    public void setPermissions(List<PermissionModel> permissionModelList) {
        this.permissionList = permissionModelList;
        this.notifyDataSetChanged();
    }
    //endregion


    //region: ViewHolder
    class PermissionViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private final String TAG = PermissionViewHolder.class.getSimpleName();
        private final PermissionRecyclerViewListener recyclerViewListener;

        private PermissionItemView permissionItemView;

        public PermissionViewHolder(@NonNull View itemView, PermissionRecyclerViewListener recyclerViewListener) {
            super(itemView);
            this.recyclerViewListener = recyclerViewListener;
            this.permissionItemView = (PermissionItemView) itemView;
            this.permissionItemView.setOnClickListener(this);
        }

        public void bindModel(PermissionModel permissionModel) {
            this.permissionItemView.setModel(permissionModel);
        }

        @Override
        public void onClick(View view) {
            this.recyclerViewListener.onSwitchStatusChanged(permissionItemView, getLayoutPosition());
        }
    }
    //endregion

}

File: app/src/main/java/com/healios/dreams/ui/permissions/PermissionRecyclerViewAdapter.kt
package com.healios.dreams.ui.permissions

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.healios.dreams.model.PermissionModel

class PermissionRecyclerViewAdapter(private var listener: PermissionRecyclerViewListener) :
    RecyclerView.Adapter<PermissionRecyclerViewAdapter.PermissionViewHolder>() {

    private var permissionList: List<PermissionModel>? = null


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PermissionViewHolder {
        return PermissionViewHolder(PermissionItemView(parent.context), listener)
    }

    override fun getItemCount(): Int {
        if (permissionList != null) {
            return this.permissionList!!.size
        }
        return 0
    }

    override fun onBindViewHolder(holder: PermissionViewHolder, position: Int) {
        val permissionModel = permissionList!![position]
        holder.bindModel(permissionModel)
    }

    //region: Setters
    fun setPermissions(permissionModelList: List<PermissionModel>) {
        this.permissionList = permissionModelList
        this.notifyDataSetChanged()
    }
    //endregion


    //region: ViewHolder
    inner class PermissionViewHolder(
        itemView: View,
        private val recyclerViewListener: PermissionRecyclerViewListener
    ) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        private val TAG = PermissionViewHolder::class.java.simpleName

        private val permissionItemView: PermissionItemView = itemView as PermissionItemView

        init {
            this.permissionItemView.setOnClickListener(this)
        }

        fun bindModel(permissionModel: PermissionModel) {
            this.permissionItemView.setModel(permissionModel)
        }

        override fun onClick(view: View) {
            this.recyclerViewListener.onSwitchStatusChanged(permissionItemView, layoutPosition)
        }
    }
    //endregion

}